R/dates.R

```diff
@@ -131,3 +160,49 @@ get_month_from_date <- function(date) {
 get_day_from_date <- function(date) {
   as.integer(as.POSIXlt(date)$mday)
 }
+
+# Simply using as.Date on a time converts based on UTC.
+# This conversion just truncates the time part.
+time_to_date <- function(date_time) 
+  as.Date(date_time, attr(date_time, "tzone"))
+
+#'Calcuate number (or proportion) of valid days in each year and quarter
+#'
+#' @importFrom lubridate quarter
+#' @importFrom tidyr spread_
+#' @param data the input data.frame
+#' @param date the date
+#' @param by the grouping variable
+#' @param units either "prop" if you want proportion of valid days or "days" is absolute number of days is required.
+#' @return data.frame with valid days in each year and quarter
+valid_by_quarter <- function(data, date, by, units = c("prop", "days")) {
+  date <- as.name(date)
+  data <- mutate(data, 
```

Comment:
I think you meant:

```
data <- mutate_(data, 
                 year    = interp(~get_year_from_date(x), x = as.name(date)), 
                 quarter = interp(~quarter(x), x = as.name(date)))
```
which is, I think, functionally the same as what I have written. I would argue, that it is a little nicer to read what I wrote, but that is certainly debatable. If you want, I can switch to the `interp` style.

I spent a lot time on deciding how I should handle NSE and SE in writing this code, and nothing felt quite right. I would love to hear your opinions on this. 

Consider the two functions that are both intended to take a data frame, and a `character` that is the name of one of its date columns, and returns the same data frame with an additional column that calculates the year of that date column:

```
# Works.
f2 <- function(data, date) {
  mutate_(data, year = interp(~rcaaqs:::get_year_from_date(x), 
                              x = as.name(date)))
}

# Also works. But perhaps cleaner?
f3 <- function(data, date) {
  date <- as.name(date)
  mutate(data, year = rcaaqs:::get_year_from_date(date))
}

df <- data.frame(a = as.Date("2000-01-01") + 365 * 1:3)
f2(df, "a") # Works
f3(df, "a") # Works

```

I would argue that `f3` is cleaner because it does not require `interp` or using `~` to make a formula, both of which I find unintuitive. Both require the use of `as.name` which is ugly, but unavoidable.

Notably, the way you had written it above fails:

```
# Fails.
f <- function(data, date) {
  mutate_(data, year = interp(~rcaaqs:::get_year_from_date(x), 
                              x = date))
}
```